transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java

```diff
@@ -71,11 +86,82 @@ public ChannelFuture register(final Channel channel, final ChannelPromise promis
         return promise;
     }
 
+    /**
+     * Adds a task to be run once at the end of next (or current) {@code eventloop} iteration.
+     *
+     * @param task to be added.
+     */
+    final void onEventLoopIteration(Runnable task) {
+        ObjectUtil.checkNotNull(task, "task");
+        if (isShutdown() || isShuttingDown() || isTerminated()) {
+            reject();
+        }
+
+        if (!tailTasks.offer(task)) {
+            throw new RejectedExecutionException("Task queue for onEventLoopIteration tasks is full.");
+        }
+
+        if (wakesUpForTask(task)) {
+            wakeup(inEventLoop());
+        }
+    }
+
+    /**
+     * Removes a task that was added previously via {@link #onEventLoopIteration(Runnable)}.
+     *
+     * @param task to be removed.
+     *
+     * @return {@code true} if the task was removed as a result of this call.
+     */
+    final boolean removeOnEventLoopIterationTask(Runnable task) {
+        return tailTasks.remove(ObjectUtil.checkNotNull(task, "task"));
+    }
+
     @Override
     protected boolean wakesUpForTask(Runnable task) {
         return !(task instanceof NonWakeupRunnable);
     }
 
+    @Override
+    protected boolean runAllTasks(long timeoutNanos) {
+        assert inEventLoop();
+        runTimeout = timeoutNanos;
+        if (lastRunTimedOut) {
+            /**
+             * If the last run timedout, then run tail tasks first, otherwise, tail tasks may be starved.
+             */
+            lastRunTimedOut = false;
+            if (runAllTasksFrom(tailTasks, timeoutNanos, runTimeoutUpdater) == DeadlineRunResult.TIMED_OUT) {
+                /**
+                 * Although, it didn't run any actual tasks, returning false, will indicate that there are no tasks
+                 * available, which we are not sure of.
+                 * Note: If every run of tail tasks and normal tasks time out, then this code will alternate, betweem
+                 * running tail tasks first and at end, which is fair. This is the reason, lastRunTimedOut is not set
+                 * to true here.
+                 */
+                return true;
+            }
+        }
+
+        return super.runAllTasks(runTimeout);
```

Comment:
> Having said that, this is also wrong because aLong is absolute

Right duration vs. time stamp issue again :) ... fwiw this means `runTimeout` could be negative :)

An option to fix this would e to change `runTimeoutUpdater` from a `LongConsumer` to a `LongBiConsumer` (which has a `void accept(long a, long b);`).  You can pass in the start time and the finish time ... this will allow you to calculate the duration and subtract it from `runTimeout` (e.g. `runTimeout -= start - stop`) ... in which case you should also check for `runTimeout` being negative before calling `super.runAllTasks(runTimeout);`